native_words.asm

```diff
@@ -11284,14 +11338,66 @@ z_unloop:       rts
 ; ## "until"  auto  ANS core
         ; """http://forth-standard.org/standard/core/UNTIL"""
 xt_until:
-                ; Compile a 0BRANCH
+                ; The address to loop back to is on the stack.
+                ; We'd normally do a 0BRANCH but similarly to xt_then
+                ; we can optimize with a native branch if it's not too far.
+                ; So we'll generate code like this
+                ;
+                ; dest:
+                ;       ...
+                ; here:
+                ;       jsr zero_branch_runtime
+                ;       LSB MSB
+                ; here+5:
+                ;
+                ; The optimized version looks like
+                ;       jsr zero_test_runtime
+                ;       beq dest - (here + 5)
+                ;
+                ; Note we could even inline zero_test_runtime but
+                ; it reduces the distance we can branch so don't bother
+
+                jsr xt_dup
+                jsr xt_here
+                jsr xt_minus            ; stack has ( dest dest-here )
+
+                lda 1,x
+                ina
+                bne _too_far            ; MSB must be #$ff
+                lda 0,x
+                sec
+                sbc #5                  ; include the extra -5
+                bpl _too_far            ; must still be negative (no wraparound)
+
+                ; good to optimize
+                sta 0,x                 ; remember the branch offset
+
+                ; write our zero_test
+                ldy #0
+-
+                lda beq_opt,y
+                jsr cmpl_a
+                iny
+                cpy #(beq_opt_end-beq_opt-1)  ; four bytes, skipping offset
+                bne -
+                lda 0,x                 ; write the offset
+                jsr cmpl_a
+
+                inx                     ; clear the stack
+                inx
+                inx
+                inx
+                rts
+
+_too_far:
+                inx                     ; discard the offset
+                inx
+                ; Just compile a 0BRANCH ...
```

Comment:
hmm I guess we could always do it natively if we wanted with a bne +3 / jmp back pattern, tho that would increase code length by two bytes and be slower in the normal repeating case.  so you'd want to do beq back if possible, otherwise do the double thing?   that's arguably simpler to read too.  wdyt?

Reply:
It would be nice to BEQ back if possible and play the "hop over the long JMP" if not possible.  I've had to do the latter myself when the branch target was too far away.  At least here, you don't have to worry about the fact that the two options are different lengths.

Also, would it be possible to give `Z` a slightly more descriptive name?  Perhaps something like `ZCHECK` or `0CHECK` or even `ZERO_CHECK`?  It seems like a useful enough word that you should consider adding it to the ASSEMBLER-WORDLIST and the glossary (it's really only useful if it has assembly following it).  That should make it automatically display its name in the disassembler without needing a handler.

Reply:
ended up going full circle on this. for until where we're writing code at the head, it's much simpler to just inline the zero test (0TEST?) and always do the bne +3/jmp back.  you could save 3 bytes/2 cycles by sometimes optimizing to the beq back but inlining already saves an extra 1000 cycles and the compiler is way easier to understand.

```
(inlined 0test with bne/jmp)
: buword 100 begin 1- ?dup 0= until ;  ok
             ' buword        cycle_test            CYCLES:  15462 ok

(jsr 0test with beq)
: buword 100 begin 1- ?dup 0= until ;  ok
             ' buword        cycle_test            CYCLES:  16463 ok
```

Reply:
Now that you've saved so many bytes/cycles, I think it's fine to spend a few in the name of simplicity.  This is still heaps faster that what Tali was doing before.

I think `0test` is a good name.  If you add it to the dictionary with neither the AlwaysNative nor NeverNative flags set, then it will either be inlined (or not) based on `nc-limit` and it should work either way, right?

One issue is that the assembly words are designed to be used in interpreted mode (otherwise you can't pass them arguments easily), and this 0test word will likely want to be used in compile mode instead.  I'm out of time today, but I'll let that rattle around in my head and see if I can figure out a good solution - it might just be good documentation on how to use it (it will need good documentation anyway, as it's not a useful word in regular Forth).  One option would be to mark it CompileOnly (CO) and you would at least get an error if you tried to use it in interpreted mode.

Reply:
The discussion about `0test` is only if you want to make it available to the user.  The other option would be to not make it available to the user and only use in internally.